File: src/main/java/com/yerti/core/command/SubCommand.java
package com.yerti.core.command;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
/**
 * Java annotation used for adding subcommands to an already existing command
 */
public @interface SubCommand {

    /**
     * REQUIRED
     * Name of the parent command
     * ex. for /fly 'fly' would be the parent for the subcommand
     * @return
     */
    String parent();

    /**
     * REQUIRED
     * Name of a command
     * Ex. /fly would have the name of 'fly'
     * @return
     */
    String name();

    /**
     * OPTIONAL
     * Permission required to issue the command
     * Leave blank for no permission
     * @return
     */
    String permission();

    /**
     * OPTIONAL, HIGHLY RECOMMENDED
     * Usage for command
     * Leave blank for no usage
     * @return
     */
    String usage();

    /**
     * OPTIONAL
     * Description for command
     * Leave blank for no description
     * @return
     */
    String description();

}

File: src/main/java/com/yerti/core/command/SubCommand.kt
package com.yerti.core.command

import java.lang.annotation.Retention
import java.lang.annotation.RetentionPolicy

@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@Retention(RetentionPolicy.RUNTIME)
/**
 * Java annotation used for adding subcommands to an already existing command
 */
annotation class SubCommand(
        /**
         * REQUIRED
         * Name of the parent command
         * ex. for /fly 'fly' would be the parent for the subcommand
         * @return
         */
        val parent: String,
        /**
         * REQUIRED
         * Name of a command
         * Ex. /fly would have the name of 'fly'
         * @return
         */
        val name: String,
        /**
         * OPTIONAL
         * Permission required to issue the command
         * Leave blank for no permission
         * @return
         */
        val permission: String,
        /**
         * OPTIONAL, HIGHLY RECOMMENDED
         * Usage for command
         * Leave blank for no usage
         * @return
         */
        val usage: String,
        /**
         * OPTIONAL
         * Description for command
         * Leave blank for no description
         * @return
         */
        val description: String)